### ISSUE TITLE
Multi Cluster support

### ISSUE THREAD

**Issue by alexeykazakov** [openshiftio/openshift.io#948]
Users can be provisioned to different OS clusters. So, OSIO services should handle it properly.

There are two main scenarios:

1. A user (space owner/collaborator) is working on a space/codebase in the same cluster.
2. A user (space collaborator) is working on a space/codebase which was created by another user and may be in a different cluster.

We will need to provide API which can be used to get the information about user's cluster. This info should be taken into account when working with a space (space owner and space collaborators may be in different clusters).

**Comment by alexeykazakov** [comment]
We need every OSIO team to prepare requirements for their services/parts. What info (provided by some new API) they would need to get to successfully support multi cluster. Any limitations and gotchas?

**Comment by alexeykazakov** [comment]
Tasks to be done to get Multi cluster support:

- [ ] Information about user's cluster should be available on the OSIO side. As first step the OSIO/OSO registration app can pass this info to the Auth service when user is approved and Auth passes it along to the tenant service. This workflow can change later as part of our registration/provisioning workflow refactoring.

**Steps to be done until we have OSO Proxy in place:**

- [ ] Move Account Linking from Keycloak to Auth. New REST API to link accounts and to get OSO (and GitHub) tokens. We will be able to get rid of OSO account linking as soon as we implement OSO proxy.
- [ ] All services/clients which talk directly to OSO should switch from Keycloak API to Auth API to get OSO tokens. UI. Che... 

**OSO Proxy:**
- [ ] OSO service accounts (one per cluster) used by OSIO should have admin permissions.
- [ ] OSO proxy implementation which uses the OSO service accounts and impersonates OSO users.
- [ ] All OSIO services/clients talk to that Proxy instead of accessing OSO directly. Che, UI, ...

**Comment by aslakknutsen** [comment]
I've noted down these as currently missing for the complete flow between registration, auth and tenant. (they might overlap with some above)

- [ ] the OS registration app doesn't send us the info
- [ ] auth can't receive the info
- [ ] auth doesn't send it to tenant service 
- [ ] the flow should change from tenant init on user registration/approval instead of first access
- [ ] tenant can't receive the cluster info
- [ ] auth can't store the cluster wide tokens
- [ ] for 5 to happen we need to get sudoers rights from OS 
- [ ] tenant service can't get the cluster tokens from auth 

@alexeykazakov Could you get @xcoulon up to speed on these and get the basic plubming for the fow up and running?

**Comment by alexeykazakov** [comment]
Sure @aslakknutsen. And I will update the list above with links to the issues we are currently working on.

**Comment by adisakala** [comment]
@surajssd

**Comment by sbose78** [comment]
>  the flow should change to tenant init on user registration/approval instead of first access

@alexeykazakov Is this something that we need to do on the Auth side after user creation is triggered from online-registration ?

**Comment by alexeykazakov** [comment]
@sbose78 yes. We should do it as soon as the registration app switches to the new API. I've created an issue for that - https://github.com/fabric8-services/fabric8-auth/issues/226

**Comment by joshuawilson** [comment]
@maxandersen SEVerity is only for bugs.

**Comment by arunkumars08** [comment]
@alexeykazakov Is the proxy url ready and if so, is there a place where I can go check on how to consume them? Just curious to know.

**Comment by sbose78** [comment]
No @arunkumars08 , The proxy url is not available yet for consumption.
@aslakknutsen is working on it.

**Comment by arunkumars08** [comment]
@sbose78 Thanks Shoubik for the update. Will follow up with @aslakknutsen

**Comment by fche** [comment]
For OSO tenant monitoring, we will only need https://github.com/fabric8-services/fabric8-tenant/issues/369 - the tenant enumeration API that gives us the cluster+tenant route base URLs.

**Comment by joshuawilson** [comment]
@alexeykazakov @aslakknutsen what are we missing to close this?

**Comment by sbose78** [comment]
"New Cluster(s) available and ready to get new users provisioned."

The above piece is not done - new cluster is available but not ready to provision users because of the performance issues being tracked by @jfchevrette .
